File: Recursion-Backtracking/06-Subset-Sum/subset-sum.cpp
class Solution {
  public:
    
    void findSums(int index, int n, int curr_sum, vector<int>& arr, vector<int>& result){
        if(index == n){
            result.push_back(curr_sum);
            return;
        }
        
        findSums(index + 1, n, curr_sum, arr, result);
        curr_sum += arr[index];
        findSums(index + 1, n, curr_sum, arr, result);
        curr_sum -= arr[index];
    }
  
    vector<int> subsetSums(vector<int>& arr) {
       vector<int> result;
       
       findSums(0, arr.size(), 0, arr, result);
       
       return result;
    }
};

File: Recursion-Backtracking/06-Subset-Sum/subset-sum.js
// User function Template for javascript

/**
 * @param {number[]} arr
 * @return {number[]}
 */

class Solution {
    
    findSums(index, n, curr_sum, arr, result, curr){
        if(index === n){
            result.push(curr_sum);
            return;
        }
        
        this.findSums(index + 1, n, curr_sum, arr, result);
        curr_sum += arr[index];
        this.findSums(index + 1, n, curr_sum, arr, result);
        curr_sum -= arr[index];
    }
    
    subsetSums(arr) {
        const result = [];
        const curr = [];
        this.findSums(0, arr.length, 0, arr, result);
        return result;
    }
}
